File: ghidra_scripts/java/inject_annotations.java
// Script Ghidra - Reinjection des annotations LLM
// Pas de dependance externe (Gson) - parsing JSON manuel
// @category LLM_Pipeline
// @author Memoire M2

import ghidra.app.script.GhidraScript;
import ghidra.app.cmd.comments.SetCommentCmd;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.SourceType;
import ghidra.program.model.address.*;
import ghidra.program.model.data.*;

import java.io.File;
import java.nio.file.Files;
import java.util.*;

public class inject_annotations extends GhidraScript {

    private int appliedCount = 0;
    private int errorCount = 0;

    // ===== Parsing JSON minimal =====

    private int pos;
    private String json;

    private void skipWhitespace() {
        while (pos < json.length() && Character.isWhitespace(json.charAt(pos))) pos++;
    }

    private char peek() {
        skipWhitespace();
        return pos < json.length() ? json.charAt(pos) : 0;
    }

    private char next() {
        skipWhitespace();
        return pos < json.length() ? json.charAt(pos++) : 0;
    }

    private String parseString() {
        if (next() != '"') return null;
        StringBuilder sb = new StringBuilder();
        while (pos < json.length()) {
            char c = json.charAt(pos++);
            if (c == '"') return sb.toString();
            if (c == '\\' && pos < json.length()) {
                char esc = json.charAt(pos++);
                switch (esc) {
                    case '"': sb.append('"'); break;
                    case '\\': sb.append('\\'); break;
                    case 'n': sb.append('\n'); break;
                    case 'r': sb.append('\r'); break;
                    case 't': sb.append('\t'); break;
                    case '/': sb.append('/'); break;
                    case 'u':
                        if (pos + 4 <= json.length()) {
                            sb.append((char) Integer.parseInt(json.substring(pos, pos + 4), 16));
                            pos += 4;
                        }
                        break;
                    default: sb.append(esc);
                }
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    private Object parseValue() {
        char c = peek();
        if (c == '"') return parseString();
        if (c == '{') return parseObject();
        if (c == '[') return parseArray();
        if (c == 't' || c == 'f') return parseBoolean();
        if (c == 'n') { pos += 4; return null; }
        return parseNumber();
    }

    private Double parseNumber() {
        skipWhitespace();
        int start = pos;
        while (pos < json.length() && "0123456789.eE+-".indexOf(json.charAt(pos)) >= 0) pos++;
        try {
            return Double.parseDouble(json.substring(start, pos));
        } catch (Exception e) {
            return 0.0;
        }
    }

    private Boolean parseBoolean() {
        skipWhitespace();
        if (json.startsWith("true", pos)) { pos += 4; return true; }
        if (json.startsWith("false", pos)) { pos += 5; return false; }
        return false;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseObject() {
        Map<String, Object> map = new LinkedHashMap<>();
        next(); // {
        while (peek() != '}') {
            String key = parseString();
            next(); // :
            Object value = parseValue();
            map.put(key, value);
            if (peek() == ',') next();
        }
        next(); // }
        return map;
    }

    private List<Object> parseArray() {
        List<Object> list = new ArrayList<>();
        next(); // [
        while (peek() != ']') {
            list.add(parseValue());
            if (peek() == ',') next();
        }
        next(); // ]
        return list;
    }

    // ===== Logique d'injection =====

    private DataType getDataType(String typeName, DataTypeManager dtm) {
        if (typeName == null || typeName.isEmpty()) return null;

        if (typeName.endsWith("*")) {
            String baseTypeName = typeName.substring(0, typeName.length() - 1).trim();
            DataType baseDt = getDataType(baseTypeName, dtm);
            if (baseDt != null) return new PointerDataType(baseDt);
            return null;
        }

        Map<String, String> typeMap = new LinkedHashMap<>();
        typeMap.put("void", "/void");
        typeMap.put("int", "/int");
        typeMap.put("char", "/char");
        typeMap.put("short", "/short");
        typeMap.put("long", "/long");
        typeMap.put("float", "/float");
        typeMap.put("double", "/double");
        typeMap.put("uint", "/uint");
        typeMap.put("uchar", "/uchar");
        typeMap.put("ushort", "/ushort");
        typeMap.put("ulong", "/ulong");
        typeMap.put("size_t", "/size_t");
        typeMap.put("bool", "/bool");

        String path = typeMap.get(typeName.toLowerCase());
        if (path != null) return dtm.getDataType(path);
        return null;
    }

    private String getStr(Map<String, Object> map, String key) {
        Object v = map.get(key);
        return (v instanceof String) ? (String) v : null;
    }

    private double getNum(Map<String, Object> map, String key) {
        Object v = map.get(key);
        return (v instanceof Number) ? ((Number) v).doubleValue() : 0.0;
    }

    @SuppressWarnings("unchecked")
    private int applySuggestion(Function func, Map<String, Object> suggestion) {
        int successCount = 0;

        double confidence = getNum(suggestion, "confidence");
        if (confidence < 0.5) {
            println("[*] Confiance faible (" + confidence + ") pour " + func.getName() + " - ignore");
            return 0;
        }

        // Renommage
        String suggestedName = getStr(suggestion, "suggested_name");
        if (suggestedName != null && !suggestedName.isEmpty()) {
            try {
                String oldName = func.getName();
                func.setName(suggestedName, SourceType.USER_DEFINED);
                println("[+] Renomme: " + oldName + " -> " + suggestedName);
                successCount++;
            } catch (Exception e) {
                println("[!] Erreur renommage: " + e.getMessage());
            }
        }

        // Type de retour
        String returnType = getStr(suggestion, "suggested_return_type");
        if (returnType != null && !returnType.isEmpty()) {
            try {
                DataType dt = getDataType(returnType, currentProgram.getDataTypeManager());
                if (dt != null) {
                    func.setReturnType(dt, SourceType.USER_DEFINED);
                    successCount++;
                }
            } catch (Exception e) {
                println("[!] Erreur type retour: " + e.getMessage());
            }
        }

        // Parametres
        Object paramObj = suggestion.get("suggested_param_types");
        if (paramObj instanceof List) {
            try {
                List<Object> paramList = (List<Object>) paramObj;
                Parameter[] params = func.getParameters();
                DataTypeManager dtm = currentProgram.getDataTypeManager();

                for (Object item : paramList) {
                    if (!(item instanceof Map)) continue;
                    Map<String, Object> pSugg = (Map<String, Object>) item;
                    String original = getStr(pSugg, "original");
                    String newName = getStr(pSugg, "suggested_name");
                    String newType = getStr(pSugg, "suggested_type");

                    for (Parameter param : params) {
                        if (param.getName().equals(original) || param.getName().contains(original != null ? original : "")) {
                            if (newName != null && !newName.isEmpty()) {
                                param.setName(newName, SourceType.USER_DEFINED);
                            }
                            if (newType != null && !newType.isEmpty()) {
                                DataType dt = getDataType(newType, dtm);
                                if (dt != null) param.setDataType(dt, SourceType.USER_DEFINED);
                            }
                            successCount++;
                            break;
                        }
                    }
                }
            } catch (Exception e) {
                println("[!] Erreur parametres: " + e.getMessage());
            }
        }

        // Commentaires
        List<String> commentParts = new ArrayList<>();
        String comments = getStr(suggestion, "comments");
        if (comments != null && !comments.isEmpty()) commentParts.add(comments);
        String reasoning = getStr(suggestion, "reasoning");
        if (reasoning != null && !reasoning.isEmpty()) commentParts.add("Analyse: " + reasoning);

        Object apisObj = suggestion.get("detected_apis");
        if (apisObj instanceof List) {
            List<Object> apis = (List<Object>) apisObj;
            if (!apis.isEmpty()) {
                StringBuilder sb = new StringBuilder("APIs: ");
                for (int i = 0; i < apis.size(); i++) {
                    if (i > 0) sb.append(", ");
                    sb.append(apis.get(i));
                }
                commentParts.add(sb.toString());
            }
        }

        if (!commentParts.isEmpty()) {
            try {
                String fullComment = String.join("\n", commentParts);
                SetCommentCmd cmd = new SetCommentCmd(func.getEntryPoint(), CodeUnit.PLATE_COMMENT, fullComment);
                cmd.applyTo(currentProgram);
                successCount++;
            } catch (Exception e) {
                println("[!] Erreur commentaire: " + e.getMessage());
            }
        }

        return successCount;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void run() throws Exception {

        println("[*] Demarrage de l'injection des annotations LLM...");

        // Recuperer le fichier JSON des arguments
        String[] args = getScriptArgs();
        String jsonFilePath;

        if (args == null || args.length == 0) {
            File binDir = new File(currentProgram.getExecutablePath()).getParentFile();
            File parentDir = binDir.getParentFile();
            jsonFilePath = new File(new File(parentDir, "suggested_files"),
                    currentProgram.getName().replace(".exe", "") + "_suggestions.json").getAbsolutePath();
            println("[*] Recherche du fichier par defaut: " + jsonFilePath);
        } else {
            jsonFilePath = args[0];
        }

        File jsonFile = new File(jsonFilePath);
        if (!jsonFile.exists()) {
            println("[!] Fichier non trouve: " + jsonFile.getAbsolutePath());
            return;
        }

        println("[*] Chargement: " + jsonFile.getAbsolutePath());

        // Lire et parser le JSON
        json = new String(Files.readAllBytes(jsonFile.toPath()), "UTF-8");
        pos = 0;

        List<Object> suggestions = parseArray();
        println("[*] " + suggestions.size() + " suggestions chargees");

        // Appliquer les suggestions
        FunctionManager funcManager = currentProgram.getFunctionManager();

        for (Object item : suggestions) {
            if (monitor.isCancelled()) break;
            if (!(item instanceof Map)) continue;

            Map<String, Object> suggestion = (Map<String, Object>) item;
            String originalName = getStr(suggestion, "original_name");

            Function func = null;

            // Recherche par nom
            if (originalName != null && !originalName.isEmpty()) {
                FunctionIterator it = funcManager.getFunctions(true);
                while (it.hasNext()) {
                    Function f = it.next();
                    if (f.getName().equals(originalName)) {
                        func = f;
                        break;
                    }
                }
            }

            if (func != null) {
                int count = applySuggestion(func, suggestion);
                appliedCount += count;
            } else {
                println("[!] Fonction non trouvee: " + originalName);
                errorCount++;
            }
        }

        println("[+] Injection terminee: " + appliedCount + " modifications, " + errorCount + " erreurs");
    }
}

File: ghidra_scripts/python/inject_annotations.py
# Script Ghidra Headless - Reinjection des annotations LLM
# A executer via: analyzeHeadless <project_dir> <project_name> -process <binary> -postScript inject_annotations.py <json_file>
# @category LLM_Pipeline
# @author Memoire M2

from ghidra.program.model.symbol import SourceType
from ghidra.program.model.data import DataTypeManager, PointerDataType
from ghidra.app.cmd.comments import SetCommentCmd
from ghidra.program.model.listing import CodeUnit
from ghidra.util.task import ConsoleTaskMonitor
import json
import os

def get_data_type(type_name, dtm):
    """Recupere ou cree un type de donnees"""
    # Types de base
    basic_types = {
        "void": dtm.getDataType("/void"),
        "int": dtm.getDataType("/int"),
        "char": dtm.getDataType("/char"),
        "short": dtm.getDataType("/short"),
        "long": dtm.getDataType("/long"),
        "float": dtm.getDataType("/float"),
        "double": dtm.getDataType("/double"),
        "uint": dtm.getDataType("/uint"),
        "uchar": dtm.getDataType("/uchar"),
        "ushort": dtm.getDataType("/ushort"),
        "ulong": dtm.getDataType("/ulong"),
        "size_t": dtm.getDataType("/size_t"),
        "bool": dtm.getDataType("/bool"),
    }

    # Gerer les pointeurs
    if type_name.endswith('*'):
        base_type = type_name[:-1].strip()
        base_dt = get_data_type(base_type, dtm)
        if base_dt:
            return PointerDataType(base_dt)

    return basic_types.get(type_name.lower())

def apply_function_rename(func, new_name):
    """Renomme une fonction"""
    try:
        func.setName(new_name, SourceType.USER_DEFINED)
        print("[+] Renomme: {} -> {}".format(func.getName(), new_name))
        return True
    except Exception as e:
        print("[!] Erreur renommage {}: {}".format(func.getName(), e))
        return False

def apply_return_type(func, type_name):
    """Change le type de retour d'une fonction"""
    try:
        dtm = currentProgram.getDataTypeManager()
        dt = get_data_type(type_name, dtm)
        if dt:
            func.setReturnType(dt, SourceType.USER_DEFINED)
            print("[+] Type retour {}: {}".format(func.getName(), type_name))
            return True
    except Exception as e:
        print("[!] Erreur type retour {}: {}".format(func.getName(), e))
    return False

def apply_parameter_types(func, param_suggestions):
    """Applique les types et noms de parametres"""
    try:
        params = func.getParameters()
        dtm = currentProgram.getDataTypeManager()

        for suggestion in param_suggestions:
            original = suggestion.get("original", "")
            new_name = suggestion.get("suggested_name")
            new_type = suggestion.get("suggested_type")

            # Trouver le parametre correspondant
            for param in params:
                if param.getName() == original or original in param.getName():
                    if new_name:
                        param.setName(new_name, SourceType.USER_DEFINED)
                        print("[+] Param renomme: {} -> {}".format(original, new_name))

                    if new_type:
                        dt = get_data_type(new_type, dtm)
                        if dt:
                            param.setDataType(dt, SourceType.USER_DEFINED)
                            print("[+] Param type: {} = {}".format(new_name or original, new_type))
                    break

        return True
    except Exception as e:
        print("[!] Erreur parametres {}: {}".format(func.getName(), e))
    return False

def apply_comment(func, comment, comment_type="plate"):
    """Ajoute un commentaire a la fonction"""
    try:
        addr = func.getEntryPoint()

        if comment_type == "plate":
            # Commentaire de type "plate" (en-tete de fonction)
            cmd = SetCommentCmd(addr, CodeUnit.PLATE_COMMENT, comment)
        elif comment_type == "pre":
            cmd = SetCommentCmd(addr, CodeUnit.PRE_COMMENT, comment)
        else:
            cmd = SetCommentCmd(addr, CodeUnit.EOL_COMMENT, comment)

        cmd.applyTo(currentProgram)
        print("[+] Commentaire ajoute pour {}".format(func.getName()))
        return True
    except Exception as e:
        print("[!] Erreur commentaire {}: {}".format(func.getName(), e))
    return False

def apply_suggestion(func, suggestion):
    """Applique toutes les suggestions pour une fonction"""
    success_count = 0

    # Verifier le seuil de confiance
    confidence = suggestion.get("confidence", 0)
    if confidence < 0.5:
        print("[*] Confiance faible ({}) pour {} - ignore".format(confidence, func.getName()))
        return 0

    # Appliquer le renommage
    if suggestion.get("suggested_name"):
        if apply_function_rename(func, suggestion["suggested_name"]):
            success_count += 1

    # Appliquer le type de retour
    if suggestion.get("suggested_return_type"):
        if apply_return_type(func, suggestion["suggested_return_type"]):
            success_count += 1

    # Appliquer les types de parametres
    if suggestion.get("suggested_param_types"):
        if apply_parameter_types(func, suggestion["suggested_param_types"]):
            success_count += 1

    # Appliquer les commentaires
    comment_parts = []
    if suggestion.get("comments"):
        comment_parts.append(suggestion["comments"])
    if suggestion.get("detected_apis"):
        apis = ", ".join(suggestion["detected_apis"])
        comment_parts.append("APIs detectees: " + apis)
    if suggestion.get("reasoning"):
        comment_parts.append("Analyse LLM: " + suggestion["reasoning"])

    if comment_parts:
        full_comment = "\n".join(comment_parts)
        if apply_comment(func, full_comment, "plate"):
            success_count += 1

    return success_count

def main():
    print("[*] Demarrage de l'injection des annotations LLM...")

    # Recuperer le fichier JSON des arguments
    args = getScriptArgs()
    if not args:
        # Chercher le fichier par defaut
        json_file = currentProgram.getName() + "_llm_suggestions.json"
        print("[*] Recherche du fichier par defaut: " + json_file)
    else:
        json_file = args[0]

    # Charger les suggestions
    if not os.path.exists(json_file):
        print("[!] Fichier non trouve: " + json_file)
        return

    with open(json_file, 'r') as f:
        suggestions = json.load(f)

    print("[*] {} suggestions chargees".format(len(suggestions)))

    # Appliquer les suggestions
    func_manager = currentProgram.getFunctionManager()
    applied_count = 0
    error_count = 0

    for suggestion in suggestions:
        address_str = suggestion.get("address", "")
        original_name = suggestion.get("original_name", "")

        # Trouver la fonction
        func = None

        # Par adresse
        if address_str:
            try:
                addr = currentProgram.getAddressFactory().getAddress(address_str)
                func = func_manager.getFunctionAt(addr)
            except:
                pass

        # Par nom
        if not func and original_name:
            functions = list(func_manager.getFunctions(True))
            for f in functions:
                if f.getName() == original_name:
                    func = f
                    break

        if func:
            count = apply_suggestion(func, suggestion)
            applied_count += count
        else:
            print("[!] Fonction non trouvee: {} / {}".format(original_name, address_str))
            error_count += 1

    print("\n[+] Injection terminee:")
    print("    {} modifications appliquees".format(applied_count))
    print("    {} erreurs".format(error_count))

if __name__ == "__main__":
    main()
